File: commonlib/src/main/java/com/sixiangtianxia/commonlib/encrypt/XXTEA.java
package com.sixiangtianxia.commonlib.encrypt;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * @author ZhangBo
 */
public class XXTEA {
    private static String KEY = "[zHanG!Can#Mou%)";

    public static String encrypt64(String value) {
        String enVid = "";
        byte[] v = value.getBytes();
        enVid = Base64.encode(XXTEA.encrypt(Base64.encode(v).getBytes(), KEY.getBytes()));
        enVid = enVid.replace('+', '-');
        enVid = enVid.replace('/', '_');
        enVid = enVid.replace('=', '.');
        return enVid;
    }

    public static String encrypt(String value) {
        String enVid = "";
        byte[] v = value.getBytes();
        enVid = Base64.encode(XXTEA.encrypt(v, KEY.getBytes()));
        enVid = enVid.replace('+', '-');
        enVid = enVid.replace('/', '_');
        enVid = enVid.replace('=', '.');
        return enVid;
    }

    public static String encrypt(HashMap<String, String> mapData) {
        StringBuffer stringBuffer = new StringBuffer();
        Iterator<Map.Entry<String, String>> iterator = mapData.entrySet().iterator();
        while (iterator.hasNext()) {
            Map.Entry<String, String> entry = iterator.next();
            stringBuffer.append(entry.getKey() + "=" + entry.getValue() + "&");
        }
        String sign = stringBuffer.substring(0, stringBuffer.length() - 1);
        return encrypt(sign);
    }

    /**
     * 先Base64解密，后 XXXTEA 解密
     *
     * @param value
     * @return
     */
    public static String decrypt(String value) {
        String deVid;
        value = value.replace('-', '+');
        value = value.replace('_', '/');
        value = value.replace('.', '=');
        byte[] v = Base64.decode(value);
        deVid = new String(XXTEA.decrypt(v));
        return deVid;
    }

    public static String decrypt64(String value) {
        String deVid;
        value = value.replace('-', '+');
        value = value.replace('_', '/');
        value = value.replace('.', '=');
        byte[] v = Base64.decode(value);
        deVid = new String(XXTEA.decrypt(v));
        return new String(Base64.decode(deVid));
    }

    /**
     * Encrypt data with key.
     *
     * @param data
     * @param key
     * @return
     */
    public static byte[] encrypt(byte[] data, byte[] key) {
        if (data.length == 0) {
            return data;
        }
        return toByteArray(encrypt(toIntArray(data, true), toIntArray(key, false)), false);
    }

    /**
     * Decrypt data with key.
     *
     * @param data
     * @return
     */
    public static byte[] decrypt(byte[] data) {
        if (data.length == 0) {
            return data;
        }
        return toByteArray(decrypt(toIntArray(data, false),
                toIntArray(KEY.getBytes(), false)), true);
    }

    /**
     * Encrypt data with key.
     *
     * @param v
     * @param k
     * @return
     */
    public static int[] encrypt(int[] v, int[] k) {
        int n = v.length - 1;
        if (n < 1) {
            return v;
        }
        if (k.length < 4) {
            int[] key = new int[4];

            System.arraycopy(k, 0, key, 0, k.length);
            k = key;
        }
        int z = v[n], y = v[0], delta = 0x9E3779B9, sum = 0, e;
        int p, q = 6 + 52 / (n + 1);
        while (q-- > 0) {
            sum = sum + delta;
            e = sum >>> 2 & 3;
            for (p = 0; p < n; p++) {
                y = v[p + 1];
                z = v[p] += (z >>> 5 ^ y << 2) + (y >>> 3 ^ z << 4) ^ (sum ^ y)
                        + (k[p & 3 ^ e] ^ z);
            }
            y = v[0];
            z = v[n] += (z >>> 5 ^ y << 2) + (y >>> 3 ^ z << 4) ^ (sum ^ y)
                    + (k[p & 3 ^ e] ^ z);
        }
        return v;
    }

    /**
     * Decrypt data with key.
     *
     * @param v
     * @param k
     * @return
     */
    public static int[] decrypt(int[] v, int[] k) {
        int n = v.length - 1;

        if (n < 1) {
            return v;
        }
        if (k.length < 4) {
            int[] key = new int[4];

            System.arraycopy(k, 0, key, 0, k.length);
            k = key;
        }
        int z = v[n], y = v[0], delta = 0x9E3779B9, sum, e;
        int p, q = 6 + 52 / (n + 1);

        sum = q * delta;
        while (sum != 0) {
            e = sum >>> 2 & 3;
            for (p = n; p > 0; p--) {
                z = v[p - 1];
                y = v[p] -= (z >>> 5 ^ y << 2) + (y >>> 3 ^ z << 4) ^ (sum ^ y)
                        + (k[p & 3 ^ e] ^ z);
            }
            z = v[n];
            y = v[0] -= (z >>> 5 ^ y << 2) + (y >>> 3 ^ z << 4) ^ (sum ^ y)
                    + (k[p & 3 ^ e] ^ z);
            sum = sum - delta;
        }
        return v;
    }

    /**
     * Convert byte array to int array.
     *
     * @param data
     * @param includeLength
     * @return
     */
    private static int[] toIntArray(byte[] data, boolean includeLength) {
        int n = (((data.length & 3) == 0) ? (data.length >>> 2)
                : ((data.length >>> 2) + 1));
        int[] result;

        if (includeLength) {
            result = new int[n + 1];
            result[n] = data.length;
        } else {
            result = new int[n];
        }
        n = data.length;
        for (int i = 0; i < n; i++) {
            result[i >>> 2] |= (0x000000ff & data[i]) << ((i & 3) << 3);
        }
        return result;
    }

    /**
     * Convert int array to byte array.
     *
     * @param data
     * @param includeLength
     * @return
     */
    private static byte[] toByteArray(int[] data, boolean includeLength) {
        byte[] result = new byte[0];
        try {
            int n = data.length << 2;
            if (includeLength) {
                int m = data[data.length - 1];
                if (m > n) {
                    return null;
                } else {
                    n = m;
                }
            }
            result = new byte[n];

            for (int i = 0; i < n; i++) {
                result[i] = (byte) ((data[i >>> 2] >>> ((i & 3) << 3)) & 0xff);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return result;
    }
}

File: commonlib/src/main/java/com/sixiangtianxia/commonlib/encrypt/XXTEA.kt
package com.sixiangtianxia.commonlib.encrypt

import java.util.HashMap

object XXTEA {
    private val KEY = "[zHanG!Can#Mou%)"

    fun encrypt64(value: String): String {
        var enVid = ""
        val v = value.toByteArray()
        enVid = Base64.encode(XXTEA.encrypt(Base64.encode(v).toByteArray(), KEY.toByteArray())!!)
        enVid = enVid.replace('+', '-')
        enVid = enVid.replace('/', '_')
        enVid = enVid.replace('=', '.')
        return enVid
    }

    fun encrypt(value: String): String {
        var enVid = ""
        val v = value.toByteArray()
        enVid = Base64.encode(XXTEA.encrypt(v, KEY.toByteArray())!!)
        enVid = enVid.replace('+', '-')
        enVid = enVid.replace('/', '_')
        enVid = enVid.replace('=', '.')
        return enVid
    }

    fun encrypt(mapData: HashMap<String, String>): String {
        val stringBuffer = StringBuffer()
        val iterator = mapData.entries.iterator()
        while (iterator.hasNext()) {
            val entry = iterator.next()
            stringBuffer.append(entry.key + "=" + entry.value + "&")
        }
        val sign = stringBuffer.substring(0, stringBuffer.length - 1)
        return encrypt(sign)
    }

    /**
     * 先Base64解密，后 XXXTEA 解密
     *
     * @param value
     * @return
     */
    fun decrypt(value: String): String {
        var value = value
        val deVid: String
        value = value.replace('-', '+')
        value = value.replace('_', '/')
        value = value.replace('.', '=')
        val v = Base64.decode(value)
        deVid = String(XXTEA.decrypt(v)!!)
        return deVid
    }

    fun decrypt64(value: String): String {
        var value = value
        val deVid: String
        value = value.replace('-', '+')
        value = value.replace('_', '/')
        value = value.replace('.', '=')
        val v = Base64.decode(value)
        deVid = String(XXTEA.decrypt(v)!!)
        return String(Base64.decode(deVid))
    }

    /**
     * Encrypt data with key.
     *
     * @param data
     * @param key
     * @return
     */
    fun encrypt(data: ByteArray, key: ByteArray): ByteArray? {
        return if (data.size == 0) {
            data
        } else toByteArray(encrypt(toIntArray(data, true), toIntArray(key, false)), false)
    }

    /**
     * Decrypt data with key.
     *
     * @param data
     * @return
     */
    fun decrypt(data: ByteArray): ByteArray? {
        return if (data.size == 0) {
            data
        } else toByteArray(
            decrypt(
                toIntArray(data, false),
                toIntArray(KEY.toByteArray(), false)
            ), true
        )
    }

    /**
     * Encrypt data with key.
     *
     * @param v
     * @param k
     * @return
     */
    fun encrypt(v: IntArray, k: IntArray): IntArray {
        var k = k
        val n = v.size - 1
        if (n < 1) {
            return v
        }
        if (k.size < 4) {
            val key = IntArray(4)

            System.arraycopy(k, 0, key, 0, k.size)
            k = key
        }
        var z = v[n]
        var y = v[0]
        val delta = -0x61c88647
        var sum = 0
        var e: Int
        var p: Int
        var q = 6 + 52 / (n + 1)
        while (q-- > 0) {
            sum = sum + delta
            e = sum.ushr(2) and 3
            p = 0
            while (p < n) {
                y = v[p + 1]
                v[p] += (z.ushr(5) xor (y shl 2)) + (y.ushr(3) xor (z shl 4)) xor (sum xor y) + (k[p and 3 xor e] xor z)
                z = v[p]
                p++
            }
            y = v[0]
            v[n] += (z.ushr(5) xor (y shl 2)) + (y.ushr(3) xor (z shl 4)) xor (sum xor y) + (k[p and 3 xor e] xor z)
            z = v[n]
        }
        return v
    }

    /**
     * Decrypt data with key.
     *
     * @param v
     * @param k
     * @return
     */
    fun decrypt(v: IntArray, k: IntArray): IntArray {
        var k = k
        val n = v.size - 1

        if (n < 1) {
            return v
        }
        if (k.size < 4) {
            val key = IntArray(4)

            System.arraycopy(k, 0, key, 0, k.size)
            k = key
        }
        var z = v[n]
        var y = v[0]
        val delta = -0x61c88647
        var sum: Int
        var e: Int
        var p: Int
        val q = 6 + 52 / (n + 1)

        sum = q * delta
        while (sum != 0) {
            e = sum.ushr(2) and 3
            p = n
            while (p > 0) {
                z = v[p - 1]
                v[p] -= (z.ushr(5) xor (y shl 2)) + (y.ushr(3) xor (z shl 4)) xor (sum xor y) + (k[p and 3 xor e] xor z)
                y = v[p]
                p--
            }
            z = v[n]
            v[0] -= (z.ushr(5) xor (y shl 2)) + (y.ushr(3) xor (z shl 4)) xor (sum xor y) + (k[p and 3 xor e] xor z)
            y = v[0]
            sum = sum - delta
        }
        return v
    }

    /**
     * Convert byte array to int array.
     *
     * @param data
     * @param includeLength
     * @return
     */
    private fun toIntArray(data: ByteArray, includeLength: Boolean): IntArray {
        var n = if (data.size and 3 == 0)
            data.size.ushr(2)
        else
            data.size.ushr(2) + 1
        val result: IntArray

        if (includeLength) {
            result = IntArray(n + 1)
            result[n] = data.size
        } else {
            result = IntArray(n)
        }
        n = data.size
        for (i in 0 until n) {
            result[i.ushr(2)] = result[i.ushr(2)] or (0x000000ff and data[i].toInt() shl (i and 3 shl 3))
        }
        return result
    }

    /**
     * Convert int array to byte array.
     *
     * @param data
     * @param includeLength
     * @return
     */
    private fun toByteArray(data: IntArray, includeLength: Boolean): ByteArray? {
        var result = ByteArray(0)
        try {
            var n = data.size shl 2
            if (includeLength) {
                val m = data[data.size - 1]
                if (m > n) {
                    return null
                } else {
                    n = m
                }
            }
            result = ByteArray(n)

            for (i in 0 until n) {
                result[i] = (data[i.ushr(2)].ushr(i and 3 shl 3) and 0xff).toByte()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return result
    }
}
